oauthproxy.go

```diff
@@ -796,6 +807,113 @@ func (p *OAuthProxy) OAuthCallback(rw http.ResponseWriter, req *http.Request) {
+//  - If no assertions were specified, it will trivially accept any (or no) claims.
+//  - Otherwise, the first assertion that evaluates to a ["truthy"](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) result will return true.
+//  - If no assertion matches, the claims are not deemed acceptable and false is returned.
+func (p *OAuthProxy) ValidateAuthorizedClaims(s *sessionsapi.SessionState, saveSession *bool) (authorized bool, reason string) {
+
+	if p.claimsAuthorizer.IsEmpty() {
+		return true, "*"
+	}
+
+	if s.Authz != "" {
+		// We have a cached claims assertions result so we only need to re-validate
+		// the request if the digest doesn't match.
+
+		// If any of those things changes, the cache is considered invalid and the full
+		// authorization check must be made again.
+
+		if providedMAC, err := base64.StdEncoding.DecodeString(s.Authz[1:]); err == nil {
+			if expectedMAC, err := p.createAuthzDigest(s, s.Authz[0]); err == nil {
+				if hmac.Equal(providedMAC, expectedMAC) {
+					// Note: The cached result is indeed valid, now actually see if we were authorized. :)
+					return s.Authz[0] == authZPass, "(cached)"
+				}
+			}
+
+			// MAC fail (probably) means something changed since we last saw this user (or
+			// the session was just refreshed), we need to re-verify the authorization rules.
+			logger.Printf("Clearing session authz cache and revalidating claims")
+		} else {
+			logger.Printf("Clearing session authz cache because of decode error: %v", err)
+		}
+
+		s.Authz = ""
+
+		// If the code path that brought us here hasn't also provided us claims to
+		// use, then we have to fail and cause a session revalidation.
+		if !s.RawClaimsValid() {
+			return false, "authz invalidated"
+		}
+	}
+
+	authzResult := authZFail
+
+	if !s.RawClaimsValid() {
+		// If we got here, then the session was created/deserialized and the
+		// provider didn't, uh, provide the claims. ;) This is an implementation
+		// error as all providers should honor this request if they are able,
+		// even if that means setting the claims to nil.
+
+		// We'll print something at least so that the person who's set up this
+		// proxy knows about it and isn't left wondering why it doesn't work.
+		logger.Printf("error: claims-based authorization is enabled, but provider implementation doesn't support it; all requests will fail to authorize")
+	} else {
+		if ok, idx := p.claimsAuthorizer.MatchesAny(s.RawClaims()); ok {
+			authzResult = authZPass
+			reason = p.claimsAuthorizer.Rules()[idx]
+		}
+	}
+
+	if digest, err := p.createAuthzDigest(s, authzResult); err != nil {
+		// Should never happen?
+		logger.Printf("error creating authz digest: %v", err)
+		authzResult = authZFail
+		reason = "digest creation failed"
+	} else {
+		s.Authz = string(authzResult) + base64.StdEncoding.EncodeToString(digest)
+		// Flag session as needing to be saved in cookie (regardless of pass or fail).
+		*saveSession = true
+	}
```

Comment:
I think this is a change in behaviour to what we normally would have done. A session being present normally implies that the user has a valid session and should be granted access. I'm not sure of the implications of changing this behaviour

Reply:
Yeah, I think I can rework this to not cache the failed case and also set up general signature-based verification of extracted claims. Stay tuned.